# Домашнє завдання до теми "Python Basics"
В цьому ДЗ важливою умовою виконання завдань є те, що ми не імпортуємо жодних бібліотек, а робимо все на чистому Python, використовуючи лише базовий функціонал. Це необхідно аби відпрацювати роботу з базовими типами даних. А використовувати бібліотеки ми ще навчимось далі :)

# Робота з рядками

1.Запишіть у змінну `name` ваше імʼя.

In [18]:
name ='Nataliia'
print (name)

Nataliia


2.Запишіть у змінну `surname` ваше прізвище.

In [19]:
surname = 'Hordiienko'
print (surname)

Hordiienko


3.Сконкатенуйте імʼя та прізвище в одну фразу через пробіл. Запишіть результат у нову змінну `name_surname` та виведіть результат:

In [46]:
name_surname = name + ' ' + surname
print  (name_surname)

Nataliia Hordiienko


Очікувана відповідь (тільки з вашими даними): 'Hanna Pylieva' 

4.Зробіть усі літери змінної `name_surname` заглавними використовуючи вбудовану функцію Python для рядків:

In [47]:
name_surname.upper()

'NATALIIA HORDIIENKO'

Очікувана відповідь (тільки з вашими даними): 'HANNA PYLIEVA'

5.Виведіть третю літеру вашого `name_surname`. Тут уважно, треба саме третя літера. Згадайте, з якого значення починається індексування значень в Python.

In [48]:
name_surname [2]

't'

Очікувана відповідь (тільки з вашими даними): 'n'

6.Виведіть першу та останню літеру вашого прізвища.

In [49]:
len (name_surname)

19

In [50]:
name_surname [0:19]

'Nataliia Hordiienko'

In [51]:
name_surname [0] + name_surname [18]

'No'

In [52]:
name_surname [0] + ' ' + name_surname [18]

'N o'

In [111]:
name_surname [0] \t name_surname [18]

SyntaxError: unexpected character after line continuation character (1350538375.py, line 1)

####  Чому при використанні /n код не працює, як правильно написати код з використанням /n та /t. Дякую :)

Очікувана відповідь (тільки з вашими даними): ('P', 'a')

#### змінений код 

In [60]:
name_surname [0] , name_surname [-1]

('N', 'o')

7.Виведіть довжину вашого прізвища.

In [11]:
len (surname)

10

Очікувана відповідь (тільки з вашими даними): 7

8.Виведіть кількість літер `a` у столиці Мадагаскару `Antananarivo`. Важливо врахувати всі літери, як великі, так і маленькі.

In [12]:
s = "Antananarivo"
s.count('a') + s.count('A')

4

Очікувана відповідь: 4

# Робота з числами та операторами

9.Створіть змінні a, b, c та d цілочисельного типу, задайде їм значення відповідно 1, 2, 3, 4, та обчисліть з допомогою Python наступний вираз:  $$ \frac{(a+b) * c}{d} $$

In [62]:
def add_numbers(a, b, c, d):
    x1 = a + b
    y1 = x1 * c
    x2 = y1 / d
    return x2

In [63]:
add_numbers(1,2,3,4)

2.25

Очікувана відповідь: 2.25

10.На змінних, створених в попердньому завданні, обчисліть наступний вираз з допомогою Python: $$ (a^b + c^d)^2 $$

In [64]:
def add_numbers(a, b, c, d):
    x3 = a ** b
    x4 = c ** d
    x5 = x3 + x4
    x6 = x5 ** 2
    return x6

In [65]:
add_numbers(1,2,3,4)

6724

Очікувана відповідь: 6724

# Робота зі колекціями

У нас є наступний список:

In [1]:
l = [
    78, 12, 14, 27, 47, 28, 51, 46, 94, 37, 42, 20, 52, 33, 11, 11, 51,
    20, 81, 54, 86, 51, 38, 42, 49, 99, 76, 17,  9, 23, 97, 29, 29, 63,
    40, 11, 31, 76, 15, 21, 29, 75, 35,  2, 10, 71, 17, 10, 13,  7, 44,
    12, 19, 61, 13, 25, 62, 46, 10, 44, 75, 57, 20, 57,  3, 44, 16, 28,
    79, 57, 62, 70, 95,  8, 47, 82, 80, 45, 36, 44, 79, 95, 73, 42, 29,
    83, 80, 66, 56,  8, 73,  5, 99, 65, 65, 98, 17, 36, 16, 60
]

Будемо з ним працювати в наступних завданнях.

11.Виведіть 5 перших елементів списку

In [2]:
l[0:5]

[78, 12, 14, 27, 47]

Очікувана відповідь: [78, 12, 14, 27, 47]

12.Виведіть максимальний і мінімальний елемент у списку

In [3]:
max (l) 

99

In [4]:
min (l)

2

In [5]:
print('Max element ', max(l))
print('Min element ', min(l))

Max element  99
Min element  2


Очікувана відповідь: 

```
Max element  99
Max element  2
```

13.Чи є в списку елементи, які повторюються? Перевірте це програмно і з допомогою інструкцій `if ... else` виведіть на екран відповідь "Так, є елементи, які повторюються", або "Ні, всі елементи унікальні".
Важливо це зробити без використання імпортів будь-яких бібліотек, на чистому Python так би мовити.

In [6]:
if len(set(l)) == len(l):
    print("Ні, всі елементи унікальні")
else:
    print("Так, є елементи, які повторюються")

Так, є елементи, які повторюються


## чому раніше код считувало а потім перестало

14.Для кожного унікального елементу списку `l` знайдіть, скільки разів він повторюється та запишіть результат у змінну `n_repeats`. Виведіть `n_repeats` на екран.

In [7]:
my_set1 = dict()
for item in l:
    n_repeats = 0
    for repeats in l:
        if repeats == item:
            n_repeats += 1
    my_set1[item ] = n_repeats
print(my_set1, end="\n")

{78: 1, 12: 2, 14: 1, 27: 1, 47: 2, 28: 2, 51: 3, 46: 2, 94: 1, 37: 1, 42: 3, 20: 3, 52: 1, 33: 1, 11: 3, 81: 1, 54: 1, 86: 1, 38: 1, 49: 1, 99: 2, 76: 2, 17: 3, 9: 1, 23: 1, 97: 1, 29: 4, 63: 1, 40: 1, 31: 1, 15: 1, 21: 1, 75: 2, 35: 1, 2: 1, 10: 3, 71: 1, 13: 2, 7: 1, 44: 4, 19: 1, 61: 1, 25: 1, 62: 2, 57: 3, 3: 1, 16: 2, 79: 2, 70: 1, 95: 2, 8: 2, 82: 1, 80: 2, 45: 1, 36: 2, 73: 2, 83: 1, 66: 1, 56: 1, 5: 1, 65: 2, 98: 1, 60: 1}


#### якщо додати end="\n" чому не змінюється формат

Очікуваний результат (три крапки - для скорочення, у вас буде виведений повний набір унікальних елементів): 
```
{2: 1,
 3: 1,
 5: 1,
    ...
 97: 1,
 98: 1,
 99: 2}
```

15.Знайдіть три найменші унікальні елементи в списку `l`.

In [8]:
three_numbers = list (set(l))
print (three_numbers [0:3])

[2, 3, 5]


Очікувана відповідь: [2, 3, 5]


16.Використовуючи `n_repeats` скажіть, скільки разів повторюється кожен з трьох найменших елементів в списку.

In [75]:
my_set2 = dict()
for item in set (l):
    n_repeats = 0
    for repeats in set (l):
        if repeats == item:
            n_repeats += 1
    my_set2[item ] = n_repeats
print (my_set2.index(0))

{2: 1, 3: 1, 5: 1, 7: 1, 8: 1, 9: 1, 10: 1, 11: 1, 12: 1, 13: 1, 14: 1, 15: 1, 16: 1, 17: 1, 19: 1, 20: 1, 21: 1, 23: 1, 25: 1, 27: 1, 28: 1, 29: 1, 31: 1, 33: 1, 35: 1, 36: 1, 37: 1, 38: 1, 40: 1, 42: 1, 44: 1, 45: 1, 46: 1, 47: 1, 49: 1, 51: 1, 52: 1, 54: 1, 56: 1, 57: 1, 60: 1, 61: 1, 62: 1, 63: 1, 65: 1, 66: 1, 70: 1, 71: 1, 73: 1, 75: 1, 76: 1, 78: 1, 79: 1, 80: 1, 81: 1, 82: 1, 83: 1, 86: 1, 94: 1, 95: 1, 97: 1, 98: 1, 99: 1}


new_dict1 = {}
for key in my_set2:
    if key == '2'
print (new_dict1)

In [146]:
new_dict1 = {}
for key in new_dict1:
    new_dict1 = n_repeats
print (new_dict1)

{}


In [144]:
new_dict1 = {}
for key  in new_dict1.items():
    new_dict1 = n_repeats
print ({key})

{78}


Очікувана відповідь:  {2: 1, 3: 1, 5: 1}

17.Створіть новий список `l2`, який містить кожен 11-й елемент списку `l`. Виведіть `l2` на екран.  
Памʼятаємо, щоб наприклад, перший 11-и1 елемент списку має індекс 10.

In [116]:
my_list_12 = l 
my_list_12 [10]

42

In [173]:
my_list_12 = l 
my_list_12 [10:99:11]

[42, 51, 29, 2, 13, 44, 80, 66, 16]

In [171]:
 len (my_list_12)

100

In [184]:
my_list_12 = l [10:99:11]
print (my_list_12)

[42, 51, 29, 2, 13, 44, 80, 66, 16]


Очікуваний результат: `[42, 51, 29, 2, 13, 44, 80, 66, 16]`

18.Створіть новий список `l2_squared`, де кожен елемент списку `l2` є піднесеним до степіня 2. 

In [187]:
set (my_list_12)

{2, 13, 16, 29, 42, 44, 51, 66, 80}

In [190]:
l2_squared = set ()
for item in my_list_12:
    l2_squared.add(item ** 2)
print (l2_squared)

{6400, 256, 1764, 4, 4356, 841, 2601, 169, 1936}


### Якщо перетворюю на сет то змінюється порядок чисел, як це змінити

Очікуваний результат: `[1764, 2601, 841, 4, 169, 1936, 6400, 4356, 256]`

19.Знайдіть суму всіх елементів списку `l2` та виведіть її на екран.

In [199]:
sum (l2_squared)

343

20.Знайдіть середнє значення елементів списку `l2`, збережіть результат у змінну `l2_mean` та виведіть новостворену змінну та ії тип на екран.

In [203]:
l2_mean = sum (l2_squared) / len (l2_squared)
print (l2_mean)
type (l2_mean)

38.111111111111114


float

21.Створіть новий список `l2_gt_mean`, за наступним правилом: "якщо цей елемент спиcку `l2` є більше за `l2_mean`, то запишемо 1, інакше 0. Виведіть новий список на екран.

In [209]:
a=0
b=1
l2_gt_mean = []
for item in my_list_12:
    if item > l2_mean:
        l2_gt_mean.append(b)
    else:
        l2_gt_mean.append(a)
print (l2_gt_mean)
    

[1, 1, 0, 0, 0, 1, 1, 1, 0]


In [206]:
my_list_12

[42, 51, 29, 2, 13, 44, 80, 66, 16]

In [207]:
l2_mean

38.111111111111114

Очікуваний результат: `[1, 1, 0, 0, 0, 1, 1, 1, 0]`

# Ураа!! Ви завершили перше домашнє завдання!!!  
Аби його здати - залийте ноутбук у гітхаб репозиторій, який ви створили у перші дні навчання на курсі, та додайте посилання на файл, заповнишви форму https://forms.gle/EvxoiJiBTkuxFmhV7   
Так тримати! 🚀🚀🚀
